File: learn/tree/fenwick-tree/fenwick_tree.go
type FenwickTree struct {
	bits []int64
}

func NewFenwickTree(n int) *FenwickTree {
	return &FenwickTree{
		bits: make([]int64, n+1),
	}
}

func lowbit(i int) int {
	return i & -i
}

func (ft *FenwickTree) Update(i int, x int64) {
	for i < len(ft.bits) {
		ft.bits[i] += x
		i += lowbit(i)
	}
}

func (ft *FenwickTree) Query(i int) int64 {
	var sum int64 = 0
	for i > 0 {
		sum += ft.bits[i]
		i -= lowbit(i)
	}
	return sum
}

func (ft *FenwickTree) QueryRange(l, r int) int64 {
	return ft.Query(r) - ft.Query(l-1)
}

File: learn/tree/fenwick-tree/fenwick_tree.cpp
#include <vector>

using namespace std;

class FenwickTree {
private:
    vector<long long> bits;

    inline int lowbit(int i) const {
        return i & -i;
    }

public:
    FenwickTree(int n) : bits(n + 1, 0) {}

    void update(int i, long long x) {
        while (i < bits.size()) {
            bits[i] += x;
            i += lowbit(i);
        }
    }

    long long query(int i) const {
        long long sum = 0;
        while (i > 0) {
            sum += bits[i];
            i -= lowbit(i);
        }
        return sum;
    }

    long long queryRange(int l, int r) const {
        return query(r) - query(l - 1);
    }
};
